app/controllers/uploads_controller.rb

```diff
@@ -1,10 +1,60 @@
 # frozen_string_literal: true
 class UploadsController < ApplicationController
+  include Flashable
+
   def index
-    @uploads = Upload.paginate(page: params[:page]).order(:created_at)
+    @uploads = Upload.paginate(page: params[:page]).order(created_at: :desc)
+  end
+
+  def new
+    csv_type = params[:csv_type]
+
+    @upload = Upload.new(csv_type: csv_type)
+    @upload.skip_lines = defaults[csv_type || 'generic']['skip_lines']
+  end
+
+  def create
+    @upload = Upload.create(merged_params)
+    data = load_csv
 
-    respond_to do |format|
-      format.html
+    if @upload.ok
+      failed_instances = data.failed_instances
+      redirect_to @upload, alert: errors_for_alert(failed_instances), notice: message_for_notice(failed_instances)
+    else
+      render :new, alert: errors_for_alert([@upload]), notice: "Failed to upload #{original_filename}."
     end
   end
+
+  def show
+    @upload = Upload.find_by(id: params[:id])
+    redirect_to uploads_path, alert: ["Upload with id: '#{params[:id]}' not found"] unless @upload
+  end
+
+  private
+
+  def load_csv
+    return unless @upload.persisted?
+
+    file = @upload.upload_file.tempfile
+    data = @upload.csv_type.constantize.load(file, skip_lines: @upload.skip_lines.try(:to_i))
+
+    @upload.update(ok: data.present? && data.ids.present?)
+    data
+  end
+
+  def original_filename
+    @f ||= upload_params[:upload_file].try(:original_filename)
```

Comment:
I think that original_filename would only be nil if the upload_file (`ActionDispatch::Http::UploadedFile`), is itself not valid. I wouldn't know where in the system rails would place the temp uploaded file. Also, the names of the CSVs may differ on every upload for the same csv (some have date and time encoded in their names).